rbd: add capability to automatically enable read affinity (#3639)
ceph/ceph-csi: docs/deploy-rbd.md

| `--maxsnapshotsonimage`  | `450`                         | Maximum number of snapshots allowed on rbd image without flattening                                                                                                                                                                                                                  |
| `--setmetadata`          | `false`                       | Set metadata on volume                                                                                                                                                                                                                                                               |
| `--enable-read-affinity` | `false`                       | enable read affinity                                                                                                                                                                                                                                                                 |

> | `--crush-location-labels`| _empty_                       | Kubernetes node labels that determines the CRUSH location the node belongs to, separated by ','                                                                                                                                                                                      |

Reviewer: ```suggestion
| `--crush-location-labels`| _empty_                       | Kubernetes node labels that determine the CRUSH location the node belongs to, separated by ','                                                                                                                                                                                      |
```

[See the Helm chart readme for installation instructions.](../charts/ceph-csi-rbd/README.md)

## Read Affinity using crush locations for RBD volumes

> CephCSI supports mapping RBD volumes with krbd options

Reviewer 2: Ceph CSI

`"read_from_replica=localize,crush_location=type1:value1|type2:value2"` to achieve
reads from the most local serving OSD.
Refer [krbd-options](https://docs.ceph.com/en/latest/man/8/rbd/#kernel-rbd-krbd-options) for more details.

> This can be enabled by adding labels to kubernetes nodes in the format `"topology.io/zone=east"` and passing

Reviewer 3: `kubernetes` to `Kubernetes`

cmdline arguments `"--enable-read-affinity=true"` and `"--crush-location-labels=topology.io/zone,topology.io/region"`

> to CephCSI RBD daemonset pod "csi-rbdplugin" container, resulting in CephCSI adding

Reviewer 2: same here

> `"--options read_from_replica=localize,crush_location=zone:east"` krbd options during rbd map operation.

Reviewer 3: Add link/predefined expected label keys here https://kubernetes.io/docs/reference/labels-annotations-taints/

Author: > Add link/predefined expected label keys here https://kubernetes.io/docs/reference/labels-annotations-taints/

We don't have a restriction at cephcsi level to have such expectations, user can use any labels he wants.

Reviewer 3: But still, you need to have a format to follow where the exact format is documented. i can add  a label is `topology.io/zone/test=east` or any arbitrary value, if we dont have any predefined strict rules defined on the format user can set any values he like in different formats

[See the Helm chart readme for installation instructions.](../charts/ceph-csi-rbd/README.md)

## Read Affinity using crush locations for RBD volumes

Ceph CSI supports mapping RBD volumes with krbd options
`"read_from_replica=localize,crush_location=type1:value1|type2:value2"` to achieve
reads from the most local serving OSD.
Refer [krbd-options](https://docs.ceph.com/en/latest/man/8/rbd/#kernel-rbd-krbd-options)
for more details.

This can be enabled by adding labels to Kubernetes nodes in the format `"topology.io/zone=east"`
and passing cmdline arguments `"--enable-read-affinity=true"` and
`"--crush-location-labels=topology.io/zone,topology.io/region"`
to Ceph CSI RBD daemonset pod "csi-rbdplugin" container, resulting in Ceph CSI adding
`"--options read_from_replica=localize,crush_location=zone:east"` krbd options during
rbd map operation.

> If enabled, this option will be added to all RBD volumes mapped by Ceph CSI.

Reviewer: It would be nice to also note that, for this setting to be effective, the OSDs should be assigned to the same  zones, regions, etc respectively on the cluster side (in the CRUSH map).  Just declaring that a particular node belongs to a particular zone on the client side is insufficient.

Author: I've modified the first sentence to elaborate on it, 
But users should find more info on ceph krbd map options site which is linked in second sentence.

[See the Helm chart readme for installation instructions.](../charts/ceph-csi-rbd/README.md)

## Read Affinity using crush locations for RBD volumes

> Ceph CSI supports mapping RBD volumes with krbd options
> `"read_from_replica=localize,crush_location=type1:value1|type2:value2"` to
> optimize for reads from the most local serving OSD (OSD assigned in the
> same ceph CRUSH map location i.e.,region,zone,rack,etc).

Reviewer: The CRUSH location doesn't have to be the _same_, the proximity gets calculated:
```suggestion
Ceph CSI supports mapping RBD volumes with krbd options
`"read_from_replica=localize,crush_location=type1:value1|type2:value2"` to
allow serving reads from the most local OSD (according to OSD locations as
defined in the CRUSH map).
```

Refer [krbd-options](https://docs.ceph.com/en/latest/man/8/rbd/#kernel-rbd-krbd-options)
for more details.

> This can be enabled by adding labels to Kubernetes nodes in the format `"topology.io/zone=east"`
> and passing cmdline arguments `"--enable-read-affinity=true"` and
> `"--crush-location-labels=topology.io/zone,topology.io/region"`
> to Ceph CSI RBD daemonset pod "csi-rbdplugin" container, resulting in Ceph CSI adding
> `"--options read_from_replica=localize,crush_location=zone:east"` krbd options during
> rbd map operation.

Reviewer: I'd extend the example to two CRUSH locations:
```suggestion
This can be enabled by adding labels to Kubernetes nodes like
`"topology.io/region=east"` and `"topology.io/zone=east-zone1"` and
passing cmdline arguments `"--enable-read-affinity=true"` and
`"--crush-location-labels=topology.io/zone,topology.io/region"` to Ceph CSI
RBD daemonset pod "csi-rbdplugin" container, resulting in Ceph CSI adding
`"--options read_from_replica=localize,crush_location=zone:east-zone1|region:east"`
krbd options during rbd map operation.
```

[See the Helm chart readme for installation instructions.](../charts/ceph-csi-rbd/README.md)

## Read Affinity using crush locations for RBD volumes

Ceph CSI supports mapping RBD volumes with krbd options
`"read_from_replica=localize,crush_location=type1:value1|type2:value2"` to
allow serving reads from the most local OSD (according to OSD locations as
defined in the CRUSH map).
Refer [krbd-options](https://docs.ceph.com/en/latest/man/8/rbd/#kernel-rbd-krbd-options)
for more details.

This can be enabled by adding labels to Kubernetes nodes like
`"topology.io/region=east"` and `"topology.io/zone=east-zone1"` and

> passing cmdline arguments `"--enable-read-affinity=true"` and

Reviewer 2: command line
